20. 프로토콜.playground/Contents.swift

```diff
@@ -0,0 +1,494 @@
+}
+
+// 수신, 발신이 가능한 Mail 클래스
+class Mail: Sendable, Receiveable {
+    var from: Sendable {
+        return self
+}
+
+    var to: Receiveable?
+
+    func send(data: Any) {
+        guard let receiver: Receiveable = self.to else {
+            print("Mail has no receiver")
+            return
+        }
+        
+        receiver.received(data: data, from: self.from)
+    }
+    
+    func received(data: Any, from: Sendable) {
+        print("Mail received \(data) from: \(from)")
+    }
+    
+    // static 메서드이므로 상속이 불가능하다.
+    static func isSendableInstance(_ instance: Any) -> Bool {
+        if let sendableInstance: Sendable = instance as? Sendable {
+            return sendableInstance.to != nil
+        }
+        return false
+    }
+}
+// 두 Message 인스턴스를 생성합니다.
+let myPhoneMessage: Message = Message()
+let yourPhoneMessage: Message = Message()
+
+// 아직 수신받을 인스턴스가 없습니다.
+myPhoneMessage.send(data: "Hello")
+
+// Message 인스턴스는 발신과 수신이 모두 가능하므로 메세지를 주고 받을 수 있다.
+myPhoneMessage.to = yourPhoneMessage
+myPhoneMessage.send(data: "Hello")
+
+// 두 Mail 인스턴스를 생성합니다.
+let myMail: Mail = Mail()
+let yourMail: Mail = Mail()
+
+myMail.send(data: "Hi")
+
+//Mail과 Message 모두 Sendable과 Receiveable 프로토콜을 준수하므로 서로 주고 받을 수 있다.
+myMail.to = yourMail
+myMail.send(data: "Hi")
+
+myMail.to = myPhoneMessage
+myMail.send(data: "Bye")
+
+// String은 Sendable프로토콜을 준수하지 않는다.
+Message.isSendableInstance("Hello")
+
+// Mail과 Message는 Sendable 프로토콜을 준수한다.
+Message.isSendableInstance(myPhoneMessage)
+
+// yourPhoneMessage는 to 프로퍼티가 설정되지 않아서 보낼 수 없는 상태이다.
+Message.isSendableInstance(yourPhoneMessage)
+Mail.isSendableInstance(myPhoneMessage)
+Mail.isSendableInstance(myMail)*/
+
+
+// 20.3.3 가변 메서드 요구
+// 프로토콜이 어떤 타입이든 간에 인스턴스 내부의 값을 변경해야 하는 메서드를 요구하려면 프로토콜의 메서드 정의 앞에 mutating 키워드를 명시해야 한다.
+
+/*// Code. Resettable 프로토콜의 가변 메서드 요구
+protocol Resettable {
+    mutating func reset() // 인스턴스 내부의 값을 변경해야하는 메서드를 요구하기 때문에 프로토콜의 메서드 정의 앞에 mutating 키워드를 명시한다.
+}
+
+class Person: Resettable { // 클래스
+    var name: String?
+    var age: Int?
+
+    func reset() { // reset() 가변 메서드 요구
+        self.name = nil
+        self.age = nil
+    }
+}
+
+struct Point: Resettable { // 값타입
+    var x: Int = 0
+    var y: Int = 0
+    
+    mutating func reset() {
+         self.x = 0
+         self.y = 0
+    }
+}
+
+enum Direction: Resettable {
+case east, west, south, north, unknown
+
+    mutating func reset() {
+        self = Direction.unknown
+    }
+}
+
+// Resettable 프로토콜에서 가변 메서드를 요구하지 않는다면, 값 타입의 인스턴스 내부 값을 변경하는 mutating 메서드는 구현이 불가능하다.*/
+
+// 20.3.4 이니셜라이저 요구
+// 프로토콜에서 이니셜라이저를 요구하려면 메서드 요구와 마찬가지로 이니셜라이저를 정의하지만 구현은 하지 않는다.
+// Code. 프로토콜의 이니셜라이저 요구와 구조체의 이니셜라이져 요구 구현
+/*protocol Named { // Named 프로토콜 구현
+    var name: String { get }
+
+    init(name: String)
+}
+
+struct Pet: Named {
+    var name: String
+
+    init(name: String) {
+        self.name = name
+    }
+}*/
+/*
+// Code. 클래스의 이니셜라이저 요구 구현
+class Person: Named {
+    var name: String
+
+    required init(name: String) {
```

Comment:
Concentrate on 'required' keyword
If your protocol require 'initializer', your class should implement 'initializer' with **'required'** keyword.